File: back-end/service/user.service.ts
import { User } from "../model/user";
import {AuthenticationResponse, User as UserInput} from '../types'
import userDb from "../repository/user.db"
import bcrypt from 'bcrypt'
import generateJwtToken from "../util/jwt";


const getAllUsers = async(role: string): Promise<User[]> => {
    if (role == 'Admin'){
        return await userDb.getAllUsers();
    } else {
        throw new Error("You are unauthorized to perform this action.")
    }
}

const getUserByEmail = async(email: string): Promise<User> => {
    const user = await userDb.getUserByEmail(email);
    
    if (!user) {
        throw new Error("User with this email does not exist.");
    }

    return user;
}

const createUser = async({username, email, password}: UserInput): Promise<AuthenticationResponse> => {
    const existing_user = await userDb.getUserByEmail(email);

    if (existing_user) {
        throw new Error("User with this email already exists.");
    }

    const hashedPassword = await bcrypt.hash(password, 12)
    const user = new User({username, email, password: hashedPassword, role: 'User'})

    await userDb.createUser(user);

    return {
        
        token: generateJwtToken(email, user.getRole()),
        username: user.getUsername(),
        email: user.getEmail(),
        role: user.getRole(),
    
    }


}

const authenticate = async({email, password}: UserInput): Promise<AuthenticationResponse> => {
    const user = await userDb.getUserByEmail(email);
    if (!user) {
        throw new Error("Email or password are wrong.");
    }

    const isValidPassword = await bcrypt.compare(password, user.getPassword());
    if (!isValidPassword) {
        throw new Error("Email or password are wrong.")
    }

    return {
        token: generateJwtToken(email, user.getRole()),
        username: user.getUsername(),
        email: email,
        role: user.getRole()
    }
}

export default {
    getAllUsers,
    getUserByEmail,
    createUser,
    authenticate
}

File: back-end-java/src/main/java/com/subvero/subverobackend/service/UserService.java
package com.subvero.subverobackend.service;

import java.util.List;

import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import com.subvero.subverobackend.controller.dto.AuthenticationResponse;
import com.subvero.subverobackend.controller.dto.UserInput;
import com.subvero.subverobackend.exception.AlreadyExistsException;
import com.subvero.subverobackend.exception.NotFoundException;
import com.subvero.subverobackend.model.Role;
import com.subvero.subverobackend.model.User;
import com.subvero.subverobackend.repository.UserRepository;

@Service
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder, JwtService jwtService) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
    }

    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    public User getUserByEmail(String email) throws NotFoundException {
        User user = userRepository.findByEmail(email);

        if (user == null) {
            throw new NotFoundException("User with this email does not exist.");
        }

        return user;
    }

    public AuthenticationResponse createUser(UserInput userInput) {
        if (userRepository.existsByUsername(userInput.username())) {
            throw new AlreadyExistsException("This user already exists");
        }

        if (userRepository.existsByEmail(userInput.email())) {
            throw new AlreadyExistsException("This user already exists");
        }

        String hashedPassword = passwordEncoder.encode(userInput.password());
        User user = new User(userInput.username(), userInput.email(), hashedPassword, Role.User);
        userRepository.save(user);

        return new AuthenticationResponse("Authenticated successfully.",
                jwtService.generateToken(user), user.getUsername(), user.getRole());

    }

    public AuthenticationResponse authenticate(String email, String password) {
        User user = userRepository.findByEmail(email);

        if (user == null) {
            throw new NotFoundException("User does not exist.");
        }

        return new AuthenticationResponse("Authenticated successfully.", jwtService.generateToken(user),
                user.getUsername(), user.getRole());
    }

}